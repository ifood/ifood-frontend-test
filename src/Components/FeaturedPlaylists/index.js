import React from 'react'

import { Container } from './styles.js'

export default function FeaturedPlaylist(){
    return(
        <Container>
            <h1>Ol√° mundo</h1>
        </Container>
    )
}