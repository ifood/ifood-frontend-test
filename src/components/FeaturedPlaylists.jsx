import React from 'react';
import PropTypes from 'prop-types';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import { If, Then } from 'react-if';

const FeaturedPlaylists = ({ filters, playlists, handleGoToPlaylist }) => (
  <div className="container featured-playlists-container">
    <h1>Confira abaixo as playlists recomendadas para vocÃª!</h1>
    <div className="row">
      <If condition={!!playlists.length}>
        <Then>
          {
                            playlists.map(playlist => (
                              <div
                                key={playlist.id}
                                className="featured-playlist-item col-md-4 col-sm-12 "
                                onClick={() => handleGoToPlaylist(playlist.external_urls.spotify)}
                              >
                                <Card
                                  className="playlist-item"
                                >
                                  <Card.Img variant="top" src={playlist.images[0].url} />
                                  <Card.Body>
                                    <Card.Title>{ playlist.name }</Card.Title>
                                    <Button
                                      variant="secondary"
                                      className="reset-filters"
                                      onClick={handleGoToPlaylist}
                                    >
                                      Ver Playlist
                                    </Button>
                                  </Card.Body>
                                </Card>
                              </div>
                            ))
                        }
        </Then>
      </If>
      <If condition={filters.length && !playlists.length}>
        <Then>
          <h5>Nenhum resultado encontrado para esse filtro! &nbsp; =/</h5>
        </Then>
      </If>
    </div>
  </div>
);

FeaturedPlaylists.propTypes = {
  playlists: PropTypes.array.isRequired,
  handleGoToPlaylist: PropTypes.func,
};

FeaturedPlaylists.defaultProps = {
  playlists: [],
};

export default FeaturedPlaylists;

// TODO: prop types
