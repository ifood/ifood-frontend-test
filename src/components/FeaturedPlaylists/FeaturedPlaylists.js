import React from 'react';
// import styles from './featured-playlists.module.css';

const PlaylistCard = (props) => (
  <div className="col-4">
    <div className="card">
    <img className="card-img-top" src={props.imageUrl} alt={props.name} />
    <div className="card-body">
      <h5 className="card-title">{props.name}</h5>
      <p className="card-text">Total tracks: {props.tracks.total}</p>
      <a href={props.externalUrl} className="btn btn-primary">See on Spotify</a>
    </div>
  </div>
  </div>
);

const FeaturedPlaylists = ({
  featuredPlaylist = {}
}) => {
  return (
    <div className="container">
      <div className="row">
      {
        Boolean(featuredPlaylist.playlists) ? featuredPlaylist.playlists.items.map(item =>
          <PlaylistCard
            name={item.name}
            imageUrl={item.images[0].url}
            tracks={item.tracks}
            externalUrl={item.external_urls.spotify}
          />
        ) : <p>Gostaria de filtrar sua playlist? :)</p>
      }
      </div>
    </div>
  )
}

export default FeaturedPlaylists;