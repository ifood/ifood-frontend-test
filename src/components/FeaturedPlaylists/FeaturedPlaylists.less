@import '../../styles/less/_parameters';

.playlist{
  padding:15px 0px;

  &__item{
    width: 250px;
    height: 250px;
    background-size: cover;
    background-position: center;
    cursor:pointer;

    &__info{
      width: 100%;
      height: 100%;
      padding:15px;

      display:flex;
      align-items:center;
      justify-content:center;

      background-color: @card-background;
      color:transparent;
      font-weight: 600;

      &:hover{
        background-color:@card-background-hover;
        color:@white;

      -webkit-transition: background-color 300ms linear, color 800ms linear;
      -moz-transition: background-color 300ms linear, color 800ms linear;
      -o-transition: background-color 300ms linear, color 800ms linear;
      -ms-transition: background-color 300ms linear, color 800ms linear;
      transition: background-color 300ms linear, color 800ms linear;
      }
    }
  }
}
