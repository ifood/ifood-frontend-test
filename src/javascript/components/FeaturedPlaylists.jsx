import React, { Component } from "react";
import PlaylistCard from "./PlaylistCard";

export default class FeaturedPlaylists extends Component {

    renderPlaylists() {
        const { playlists } = this.props;
        console.log(playlists)

        return playlists.map(playlist => <PlaylistCard playlist={playlist} />);
    }

    render() {
        const { accessToken, fetchFeaturedPlaylists, playlists } = this.props;
        accessToken && playlists.length === 0 && fetchFeaturedPlaylists(accessToken);

        return (
            <div className="playlists-list">
                { this.renderPlaylists() }
            </div>
        )
    }
}
