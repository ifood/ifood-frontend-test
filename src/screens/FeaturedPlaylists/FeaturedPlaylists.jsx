import React, { useEffect, useState, useMemo, useCallback } from 'react';
import { debounce } from 'throttle-debounce';
import { useIntl } from 'react-intl';
import PlaylistGrid from '../../component/PlaylistGrid';
import { useAuthorization } from '../../hooks';
import Filter from '../../component/Filter';
import Info from '../../component/Info';
import {
  LayoutContainer,
  Header,
  Content,
  Footer,
} from '../../component/Layout';
import {
  getFilterMetadata,
  getFeaturedPlayList,
} from '../../network/FeaturedPlayListNetwork';
import { parseObjectByMetatada } from '../../component/DynamicField/FieldMetadata';
import Loading from '../../component/Loading';

const SCREEN_STATE = {
  LOADING: Symbol('loading'),
  DONE: Symbol('done'),
  ERROR: Symbol('error'),
};
const DEBOUNCE_DELAY = 300;
const POLLING_CYCLE = 1000 * 30;

function filterPlaylist(playlists, filterText) {
  return playlists?.items?.filter(
    (item) =>
      filterText == null ||
      item?.name?.toLowerCase().includes(filterText.trim().toLowerCase())
  );
}

function FeaturedPlaylists() {
  const authorization = useAuthorization();
  const [data, setData] = useState(null);
  const [filter, setFilter] = useState(null);
  const [screenState, setScreenState] = useState(SCREEN_STATE.LOADING);
  const [errorMessage, setError] = useState(null);
  const [filterText, setFilterText] = useState(null);
  const [filterMetadata, setFilterMetadata] = useState(null);
  const intl = useIntl();

  const handleResponseError = useMemo(
    () => (error) => {
      let errorResponse;
      switch (error?.status) {
        case 400:
        case 404:
        case 500:
        case 503:
        case 429:
          errorResponse = JSON.parse(error.response);
          setError(errorResponse?.error?.message);
          break;

        case 401:
        case 403:
          authorization.logout();
          break;

        default:
          break;
      }

      setScreenState(SCREEN_STATE.ERROR);
    },
    [authorization]
  );

  const loadData = useMemo(
    () => ({ silent = false } = {}) => {
      if (!silent) {
        setScreenState(SCREEN_STATE.LOADING);
      }
      getFeaturedPlayList(
        authorization.access_token,
        parseObjectByMetatada(filterMetadata, filter)
      )
        .then((response) => {
          setData(response);
          setScreenState(SCREEN_STATE.DONE);
        })
        .catch(handleResponseError);
    },
    // this hook shoud not depend on filterMetadata
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [authorization, filter, handleResponseError]
  );

  useEffect(() => {
    loadData();
  }, [loadData]);

  const refresh = useMemo(
    () => () => {
      loadData({ silent: true });
    },
    [loadData]
  );

  useEffect(() => {
    getFilterMetadata().then(setFilterMetadata);
  }, []);

  useEffect(() => {
    const interval = setInterval(refresh, POLLING_CYCLE);

    return () => clearInterval(interval);
  }, [refresh]);

  const handleFilter = useCallback((newFilter) => {
    setFilter(newFilter);
  }, []);

  const handleFilterTextChange = useCallback(
    debounce(DEBOUNCE_DELAY, setFilterText),
    []
  );

  const filtered = useMemo(() => filterPlaylist(data?.playlists, filterText), [
    data,
    filterText,
  ]);

  return (
    <LayoutContainer>
      <Header />
      <Content>
        <Filter
          filter={filter}
          metadata={filterMetadata}
          onFilter={handleFilter}
          onTextChange={handleFilterTextChange}
        />
        {screenState === SCREEN_STATE.LOADING && <Loading />}
        {screenState === SCREEN_STATE.DONE && (
          <PlaylistGrid title={data?.message} items={filtered} />
        )}
        {screenState === SCREEN_STATE.ERROR && (
          <Info
            icon="times"
            title={intl.formatMessage({ id: 'ops' })}
            info={errorMessage}
          />
        )}
      </Content>
      <Footer />
    </LayoutContainer>
  );
}

export default FeaturedPlaylists;
