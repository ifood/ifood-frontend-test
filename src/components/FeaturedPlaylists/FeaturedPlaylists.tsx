import React, { useCallback, useMemo, useState } from 'react';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';
import debounce from 'lodash/debounce';
import { Filter } from '../Filter/Filter';
import { Playlists } from '../Playlists/Playlists';
import { useFeaturedPlaylists, FeaturedPlaylistsParams } from '../../hooks/useFeaturedPlaylists';
import { useStyles } from './FeaturedPlaylists.jss';

export const FeaturedPlaylists: React.FC = () => {
  const classes = useStyles();
  const [search, setSearch] = useState('');
  const [params, setParams] = useState<FeaturedPlaylistsParams>({});
  const { featuredPlaylists, isLoading, isSuccess, isError } = useFeaturedPlaylists(params);

  const playlists = useMemo(() => {
    if (!featuredPlaylists) {
      return [];
    }
    if (!search) {
      return featuredPlaylists;
    }
    const normalizedSearch = search.toLowerCase();
    return featuredPlaylists.filter((featuredPlaylist) =>
      featuredPlaylist.name.toLowerCase().includes(normalizedSearch)
    );
  }, [featuredPlaylists, search]);

  const handleFilter = useCallback((filters: Record<string, string>) => {
    setParams(filters);
  }, []);

  const handleSearch = useMemo(() => {
    return debounce((text: string) => setSearch(text), 400);
  }, []);

  return (
    <>
      <Filter onFilter={handleFilter} onSearch={handleSearch} />

      {isLoading && (
        <div className={classes.indicator}>
          <CircularProgress color="primary" />
        </div>
      )}

      {isError && (
        <div className={classes.indicator}>
          <Typography>Não foi possível carregar as playlists.</Typography>
        </div>
      )}

      {isSuccess && playlists.length === 0 && (
        <div className={classes.indicator}>
          <Typography>Nenhuma playlist encontrada.</Typography>
        </div>
      )}

      {isSuccess && playlists.length > 0 && <Playlists playlists={playlists} />}
    </>
  );
};
