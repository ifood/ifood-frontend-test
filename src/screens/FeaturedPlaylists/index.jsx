import FeaturedPlaylists from './FeaturedPlaylists';

export default FeaturedPlaylists;
