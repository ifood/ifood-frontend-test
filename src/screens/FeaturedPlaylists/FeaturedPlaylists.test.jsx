import React from 'react';
import { mount } from 'enzyme';
import Axios from 'axios';
import toJson from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';

import FeaturedPlaylists from './FeaturedPlaylists';

jest.mock('../../hooks');

const get = () => {
  return new Promise((resolve) => {
    resolve({ data: { filters: [] } });
  });
};
Axios.get.mockImplementation(get);

describe('<FeaturedPlaylists />', () => {
  it('snapshot', async () => {
    const wrapper = mount(<FeaturedPlaylists />);
    await act(async () => wrapper);
    wrapper.update();
    jestExpect(toJson(wrapper)).toMatchSnapshot();
  });
});
