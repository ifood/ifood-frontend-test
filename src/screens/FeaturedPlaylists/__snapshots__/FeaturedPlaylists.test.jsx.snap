// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeaturedPlaylists /> snapshot 1`] = `
<FeaturedPlaylists>
  <styled.div>
    <div
      className="sc-fzokOt kuOybX"
    >
      <Header>
        <styled.header>
          <header
            className="sc-fznyAO Wnbwo"
          >
            <styled.h1>
              <h1
                className="sc-fznKkj bTYONO"
              >
                <styled.img
                  alt="Spotifood logo"
                  src="logo.png"
                >
                  <img
                    alt="Spotifood logo"
                    className="sc-fznZeY fDeeBe"
                    src="logo.png"
                  />
                </styled.img>
              </h1>
            </styled.h1>
          </header>
        </styled.header>
      </Header>
      <styled.main>
        <main
          className="sc-Axmtr iTXiqD"
        >
          <Memo(Filter)
            filter={null}
            metadata={Array []}
            onFilter={[Function]}
            onTextChange={[Function]}
          >
            <styled.div>
              <div
                className="sc-AxhUy eBYwie"
              >
                <Styled(Input)
                  icon={true}
                  placeholder="intl::search"
                >
                  <Input
                    className="sc-AxheI jlkZqc"
                    icon={true}
                    placeholder="intl::search"
                    type="text"
                  >
                    <div
                      className="ui icon input sc-AxheI jlkZqc"
                    >
                      <input
                        key=".0"
                        onChange={[Function]}
                        placeholder="intl::search"
                        type="text"
                      />
                      <Icon
                        as="i"
                        key=".1"
                        name="search"
                      >
                        <i
                          aria-hidden="true"
                          className="search icon"
                          onClick={[Function]}
                        />
                      </Icon>
                    </div>
                  </Input>
                </Styled(Input)>
                <styled.div>
                  <div
                    className="sc-AxgMl jCkrga"
                  >
                    <Button
                      as="button"
                      basic={true}
                      circular={true}
                      icon="filter"
                      onClick={[Function]}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <button
                              class="ui basic circular icon button"
                            >
                              <i
                                aria-hidden="true"
                                class="filter icon"
                              />
                            </button>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <button
                                class="ui basic circular icon button"
                              >
                                <i
                                  aria-hidden="true"
                                  class="filter icon"
                                />
                              </button>,
                            }
                          }
                        >
                          <button
                            className="ui basic circular icon button"
                            onClick={[Function]}
                          >
                            <Icon
                              as="i"
                              name="filter"
                            >
                              <i
                                aria-hidden="true"
                                className="filter icon"
                                onClick={[Function]}
                              />
                            </Icon>
                          </button>
                        </RefFindNode>
                      </Ref>
                    </Button>
                    <FilterModal
                      filter={null}
                      metadata={Array []}
                      onClose={[Function]}
                      onSubmit={[Function]}
                      open={false}
                    >
                      <Modal
                        centered={false}
                        closeOnDimmerClick={true}
                        closeOnDocumentClick={false}
                        dimmer={true}
                        eventPool="Modal"
                        onClose={[Function]}
                        open={false}
                        size="tiny"
                      >
                        <Portal
                          closeOnDocumentClick={false}
                          closeOnEscape={true}
                          eventPool="Modal"
                          mountNode={<body />}
                          onClose={[Function]}
                          onMount={[Function]}
                          onOpen={[Function]}
                          onUnmount={[Function]}
                          open={false}
                          openOnTriggerClick={true}
                        />
                      </Modal>
                    </FilterModal>
                  </div>
                </styled.div>
              </div>
            </styled.div>
            <TagList
              formatter={[Function]}
              items={null}
              onDeleteTag={[Function]}
              title="intl::filters"
            />
          </Memo(Filter)>
          <Loading>
            <Dimmer
              active={true}
              inverted={true}
            >
              <DimmerInner
                active={true}
                inverted={true}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <div
                        class="ui active transition visible inverted dimmer"
                        style="display: flex;"
                      >
                        <div
                          class="content"
                        >
                          <div
                            class="ui inverted text loader"
                          />
                        </div>
                      </div>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <div
                          class="ui active transition visible inverted dimmer"
                          style="display: flex;"
                        >
                          <div
                            class="content"
                          >
                            <div
                              class="ui inverted text loader"
                            />
                          </div>
                        </div>,
                      }
                    }
                  >
                    <div
                      className="ui active transition visible inverted dimmer"
                      onClick={[Function]}
                    >
                      <div
                        className="content"
                      >
                        <Loader
                          inverted={true}
                        >
                          <div
                            className="ui inverted text loader"
                          >
                            <FormattedMessage
                              id="loading"
                            />
                          </div>
                        </Loader>
                      </div>
                    </div>
                  </RefFindNode>
                </Ref>
              </DimmerInner>
            </Dimmer>
          </Loading>
        </main>
      </styled.main>
      <Footer>
        <styled.footer>
          <footer
            className="sc-AxmLO eyYpJA"
          >
            <styled.div>
              <div
                className="sc-fzozJi cBaOqJ"
              >
                <styled.div>
                  <div
                    className="sc-fzoLsD eCAhqh"
                  >
                    <styled.img
                      alt="Spotifood logo"
                      src="logo-icon.png"
                    >
                      <img
                        alt="Spotifood logo"
                        className="sc-fzpans kgyVqc"
                        src="logo-icon.png"
                      />
                    </styled.img>
                  </div>
                </styled.div>
                <styled.div>
                  <div
                    className="sc-fzplWN bEKLLk"
                  >
                    <p>
                      <FormattedMessage
                        id="disclaimer"
                      />
                    </p>
                    <p>
                      <FormattedMessage
                        id="disclaimer.data"
                      />
                    </p>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </footer>
        </styled.footer>
      </Footer>
    </div>
  </styled.div>
</FeaturedPlaylists>
`;
